include/seqan3/alignment/pairwise/detail/policy_optimum_tracker_simd.hpp

```diff
@@ -0,0 +1,201 @@
+        auto && [current_cell_score, current_cell_coordinate] = current_score_coordinate_pair;
+
+        auto mask = (optimal_coordinate.col == current_cell_coordinate.col) &&
+                    (optimal_coordinate.row == current_cell_coordinate.row);
+        optimal_score = (mask) ? current_cell_score : optimal_score;
+    }
+};
+
+/*!\brief Implements the tracker to store the global optimum for a particular alignment computation.
+ * \ingroup pairwise_alignment
+ * \copydetails seqan3::detail::policy_optimum_tracker
+ */
+template <typename alignment_configuration_t, std::semiregular binary_update_operation_t>
+//!\cond
+    requires is_type_specialisation_of_v<alignment_configuration_t, configuration>
+//!\endcond
+class policy_optimum_tracker_simd :
+    protected policy_optimum_tracker<alignment_configuration_t, binary_update_operation_t>
+{
+protected:
+    //!\brief The type of the base class.
+    using base_policy_t = policy_optimum_tracker<alignment_configuration_t, binary_update_operation_t>;
+
+    // Import the configured score type.
+    using typename base_policy_t::traits_type;
+    using typename base_policy_t::score_type;
+
+    //!\brief The scalar type of the simd vector.
+    using scalar_type = typename simd::simd_traits<score_type>::scalar_type;
+    //!\brief The original non-simd score type.
+    using original_score_type = typename traits_type::original_score_type;
+
+    static_assert(simd_concept<score_type>, "Must be a simd type!");
+
+    // Import base variables into class scope.
+    using base_policy_t::compare_and_set_optimum;
+    using base_policy_t::optimal_score;
+    using base_policy_t::optimal_coordinate;
+    //!\brief The individual offsets used for padding the sequences.
+    std::array<original_score_type, simd_traits<score_type>::length> padding_offsets{};
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    policy_optimum_tracker_simd() = default; //!< Defaulted.
+    policy_optimum_tracker_simd(policy_optimum_tracker_simd const &) = default; //!< Defaulted.
+    policy_optimum_tracker_simd(policy_optimum_tracker_simd &&) = default; //!< Defaulted.
+    policy_optimum_tracker_simd & operator=(policy_optimum_tracker_simd const &) = default; //!< Defaulted.
+    policy_optimum_tracker_simd & operator=(policy_optimum_tracker_simd &&) = default; //!< Defaulted.
+    ~policy_optimum_tracker_simd() = default; //!< Defaulted.
+
+    /*!\brief Construction and initialisation using the alignment configuration.
+     * \param[in] config The alignment configuration (not used in this context).
+     *
+     * \details
+     *
+     * Resets the optimum on construction. Sets track_last_row_cell and track_last_column_cell to true in order
+     * to check if this is a cell with a potential optimum.
+     */
+    policy_optimum_tracker_simd(alignment_configuration_t const & SEQAN3_DOXYGEN_ONLY(config))
+    {
+        base_policy_t::test_last_row_cell = true;
```

Comment:
> The constructor of base_policy_t has if constexpr (traits_type::with_free_end_gaps) why does this one not have this?

Because this has nothing to do with free end gaps yet. This is the semantic needed for how we implement global alignments with simd. We always look into the last row and last column and test via the coordinates whether or not this cell should be tracked. Since in the global alignment we only have one optimum whose position is fixed.

> I still think having a simd class deriving from a scalar class is wrong, because simd is more general than scalar

So you are basically saying that OOP means that the derived class has always less features than the base class? I can do everything with simd that I can do with scalar, except that comparison does not result in a bool (it is overloaded) and that I have a subscript operator to access a single element of the simd vector (scalar is extended). So in my OOP world, that would mean I extend the interface of the scalar type and thus simd is more specialised than scalar. Your example would rather be `scalar=point` and `simd=ellipse`.  -> `class ellipse : point`

Reply:
I think this is exactly the problem, they are both not complete sub-types of each other, so a normal "inheritance" does not fit. We should go by their skills and if you don't use `simd[]` and are ignoring that `simd == simd = mask != bool` (you don't need that property most of the time) and use special CPOs (like)

* `seqan3::numeric::is_mask_v<bool / simd_mask> = true`
* `seqan3::numeric::is_signed_v<int_simd_or_scalar> = true`
* `seqan3::numeric::is_unsigned_v<unsigned_simd_or_scalar> = true`
* `seqan3::numeric::make_signed_t<unsigned_simd_or_scalar>`
* `seqan3::numeric::make_unsigned_t<signed_simd_or_scalar>`
* `seqan3::numeric::assign_scalar<simd_or_scalar>(scalar) -> simd_or_scalar`,
* `seqan3::numeric::masked_assign(simd_or_scalar_mask, simd_or_scalar) -> simd_or_scalar` 
* `seqan3::numeric::ternary(simd_or_scalar_mask, simd_or_scalar, simd_or_scalar) -> simd_or_scalar` 
* `seqan3::numeric::zero<simd_or_scalar[_mask]>() -> simd_or_scalar[_mask]`
* `seqan3::numeric::one<simd_or_scalar[_mask]>() -> simd_or_scalar[_mask]`
* `seqan3::numeric::iota<simd_or_scalar>() -> simd_or_scalar`
* `seqan3::numeric::min(simd_or_scalar, simd_or_scalar) -> simd_or_scalar_mask`
* `seqan3::numeric::max(simd_or_scalar, simd_or_scalar) -> simd_or_scalar_mask`
* `seqan3::numeric::all_of(simd_or_scalar_mask) -> bool`
* `seqan3::numeric::any_of(simd_or_scalar_mask) -> bool`

I think we can express most kernels exactly the same. Of course do you need more pre- / post-processing code with simd, but I think these can live in the same class with some constexpr if's or `requires seqan3::simd::simd_concept<>` overloads.

Reply:
Well the problem is, that simd is a difficult type. It merges the concepts of arithmetic types and array-like types. The only safe abstraction would be as you suggest to handle scalar types as a special alias for simd[0]. You could have a dedicated simd_scalar_wrapper that fulfills the requirements for simd, write one kernel and cross the fingers that the compiler can optimize this away. We even talked about it sometimes in the past. But we don‘t have it. And yes you can spent another two months to make it „right“. But I don‘t have the time to do so and so has no one else. It is not planed and it does not generate sufficient value to make it a priority. If we have the feature at some point in the future, then I’ll gladly spent one afternoon to adapt the code. Right now, it is good as it is.

Reply:
I never said that this should block anything, I try to find a compromise here and I get the PR back to review just as it is, without any constructive message. The message that I received is that you don't want to change anything at all, because it is inconvenient or not "planned", and I don't know how to feel about this.

Reply:
No, it's not that I won't change it because I don't want to or because it is inconvenient. I don't change it because I haven't seen the argument yet, that choosing another way to express the same thing is beneficial.